bundles/org.openhab.binding.tuya/src/main/java/org/openhab/binding/tuya/internal/handler/TuyaDeviceHandler.java

```diff
@@ -0,0 +1,658 @@
+        if (configuration.dp2 != 0) {
+            List<String> list = Objects
+                    .requireNonNull(dp2ToChannelId.computeIfAbsent(configuration.dp2, ArrayList::new));
+            list.add(channelId);
+        }
+        if (CHANNEL_TYPE_UID_IR_CODE.equals(channelTypeUID)) {
+            irStartLearning(configuration.activeListen);
+        }
+    }
+
+    private void connectDevice() {
+        TuyaDevice tuyaDevice = this.tuyaDevice;
+        if (tuyaDevice == null) {
+            logger.warn("Cannot connect {} because the device is not set.", thing.getUID());
+            return;
+        }
+        // clear the future here because timing issues can prevent the next attempt if we fail again
+        reconnectFuture = null;
+        tuyaDevice.connect();
+    }
+
+    private List<CommandOption> toCommandOptionList(List<String> options) {
+        return options.stream().map(c -> new CommandOption(c, c)).collect(Collectors.toList());
+    }
+
+    private void addSingleExpiringCache(Integer key, Object value) {
+        ExpiringCache<@Nullable Object> expiringCache = new ExpiringCache<>(Duration.ofSeconds(10), () -> null);
+        expiringCache.putValue(value);
+        deviceStatusCache.put(key, expiringCache);
+    }
+
+    @Override
+    protected void updateState(String channelId, State state) {
+        channelStateCache.put(channelId, state);
+        super.updateState(channelId, state);
+    }
+
+    private long convertHexCode(String code) {
+        String sCode = code.startsWith("0x") ? code.substring(2) : code;
+        return Long.parseLong(sCode, 16);
+    }
+
+    private String convertBase64Code(ChannelConfiguration channelConfig, String encoded) {
+        String decoded = "";
+        try {
+            if (channelConfig.irType.equals("nec")) {
+                decoded = IrUtils.base64ToNec(encoded);
+                IrCode code = Objects.requireNonNull(gson.fromJson(decoded, IrCode.class));
+                decoded = "0x" + code.hex;
+            } else if (channelConfig.irType.equals("samsung")) {
+                decoded = IrUtils.base64ToSamsung(encoded);
+                IrCode code = Objects.requireNonNull(gson.fromJson(decoded, IrCode.class));
+                decoded = "0x" + code.hex;
+            } else {
+                if (encoded.length() > 68) {
+                    decoded = IrUtils.base64ToNec(encoded);
+                    if (decoded.isEmpty()) {
+                        decoded = IrUtils.base64ToSamsung(encoded);
+                    }
+                    IrCode code = Objects.requireNonNull(gson.fromJson(decoded, IrCode.class));
+                    decoded = code.type + ": 0x" + code.hex;
+                } else {
+                    decoded = encoded;
+                }
+            }
+        } catch (JsonSyntaxException e) {
+            logger.warn("Incorrect json response: {}", e.getMessage());
+            decoded = encoded;
+        } catch (RuntimeException e) {
```

Comment:
@J-N-K - have you ever seen this occur? Catching `RuntimeException` here can potentially hide bugs, I would assume this can only happen in case of bugs within the decoder methods?